### DISCUSSION TITLE
Unable to use tornado websocket/http client

### DISCUSSION THREAD

**Discussion by dev-dsp** [fastapi/fastapi#7155] [Questions] [answered]
### First check

* [?] I added a very descriptive title to this issue.
* [+] I used the GitHub search to find a similar issue and **found it**.
* [n/a] I searched the FastAPI documentation, with the integrated search.
* [+] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [n/a] I already read and followed all the tutorial in the docs and didn't find an answer.
* [n/a] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [n/a] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [n/a] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [+] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Description

I am trying to use fastapi with [gremlin-python](https://github.com/apache/tinkerpop/tree/master/gremlin-python/src/main/python). That library uses tornado as transport for connecting to gremlin server, especially it uses httpclient and websockets along with tornado's [ioloop](https://github.com/apache/tinkerpop/blob/master/gremlin-python/src/main/python/gremlin_python/driver/tornado/transport.py#L30).

Problem is that python prohibits nested event loops and I get runtime errors: https://pastebin.com/mcFhuKks
Currently as workaround I use [nest_asyncio](https://github.com/erdewit/nest_asyncio), but that's a very unreliable solution.

Since previously I saw several issues here regarding asyncio loops (though mostly regarding aiohttp), I am curious if there is some simple way to solve that problem (e.g. somehow reuse parent event-loop).

Also I'd like to notice that gremlin-python DB connection constructor supports passing custom transport factory (instead of [default](https://github.com/apache/tinkerpop/blob/master/gremlin-python/src/main/python/gremlin_python/driver/tornado/transport.py#L27)), and I tried to use something like `asyncio.get_event_loop()` and `asyncio.run_until_complete()` instead of tornado's alternatives, but with no luck.

Hence I'm here for any ideas or probable solutions.

### Environment

* OS: [e.g. Linux / Windows / macOS]: python:3.8-slim docker
* FastAPI Version: ==0.61.1
* Python version: python:3.8

**Comment by dev-dsp** [comment, marked as answer]
Ok, seems that I've solved this problem by using synchronous websocket library.

```
class WebsocketTransport(AbstractBaseTransport):
    def __init__(self, read_timeout=30, write_timeout=30):
        self._closed = True

    def connect(self, url, headers=None):
        if headers:
            self._ws = websocket.create_connection(url, header=headers)
        else:
            self._ws = websocket.create_connection(url)
        self._closed = False

    def write(self, message):
        self._ws.send_binary(bytes(message))

    def read(self):
        return self._ws.recv()

    def close(self):
        self._ws.close()
        self._closed = True

    def closed(self):
        return self._closed
```

Going to submit Jira issue to Tinkerpop about bad architecture of their library though.

**Comment by tiangolo** [comment]
Thanks for coming back to close the issue @dev-dsp ☕ 

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
